contracts/protocols/mainnet/uniswap/helpers.sol

```diff
@@ -0,0 +1,264 @@
+        uint256 len = nftManager.balanceOf(user);
+        tokenId = nftManager.tokenOfOwnerByIndex(user, len - 1);
+    }
+
+    function userNfts(address user) internal view returns (uint256[] memory tokenIds) {
+        uint256 len = nftManager.balanceOf(user);
+        for (uint256 i = 0; i < len; i++) {
+            uint256 tokenId = nftManager.tokenOfOwnerByIndex(user, i);
+            tokenIds[i] = tokenId;
+        }
+    }
+
+    function getMinAmount(
+        TokenInterface token,
+        uint256 amt,
+        uint256 slippage
+    ) internal view returns (uint256 minAmt) {
+        uint256 _amt18 = convertTo18(token.decimals(), amt);
+        minAmt = wmul(_amt18, sub(WAD, slippage));
+        minAmt = convert18ToDec(token.decimals(), minAmt);
+    }
+
+    function positions(uint256 tokenId)
+        internal
+        view
+        returns (
+            address token0,
+            address token1,
+            uint24 fee,
+            int24 tickLower,
+            int24 tickUpper,
+            uint128 liquidity
+        )
+    {
+        (, , token0, token1, fee, tickLower, tickUpper, liquidity, , , , ) = nftManager.positions(tokenId);
+    }
+
+    function getPoolAddress(
+        address token0,
+        address token1,
+        uint24 fee
+    ) internal view returns (address poolAddr) {
+        poolAddr = PoolAddress.computeAddress(
+            nftManager.factory(),
+            PoolAddress.PoolKey({ token0: token0, token1: token1, fee: fee })
+        );
+    }
+
+    function depositAmount(
+        uint256 tokenId,
+        uint256 amountA,
+        uint256 amountB
+    )
+        internal
+        view
+        returns (
+            uint128 liquidity,
+            uint256 amount0,
+            uint256 amount1
+        )
+    {
+        (, , address _token0, address _token1, uint24 _fee, int24 tickLower, int24 tickUpper, , , , , ) = nftManager
+        .positions(tokenId);
+
+        IUniswapV3Pool pool = IUniswapV3Pool(getPoolAddress(_token0, _token1, _fee));
+
+        // compute the liquidity amount
+        {
+            (uint160 sqrtPriceX96, , , , , , ) = pool.slot0();
+
+            liquidity = LiquidityAmounts.getLiquidityForAmounts(
+                sqrtPriceX96,
+                TickMath.getSqrtRatioAtTick(tickLower),
+                TickMath.getSqrtRatioAtTick(tickUpper),
+                amountA,
+                amountB
+            );
+
+            (amount0, amount1) = LiquidityAmounts.getAmountsForLiquidity(
+                sqrtPriceX96,
+                TickMath.getSqrtRatioAtTick(tickLower),
+                TickMath.getSqrtRatioAtTick(tickUpper),
+                liquidity
+            );
+
+            amount0 = sub(amountA, amount0);
+            amount1 = sub(amountB, amount1);
+        }
+    }
+
+    function singleDepositAmount(
+        uint256 tokenId,
+        address tokenA,
+        uint256 amountA
+    ) internal view returns (address tokenB, uint256 amountB) {
```

Comment:
@cryptoDev222 We also need to return the correct order of the token addresses and amounts.